example/32_batched_gemm_scale_softmax_gemm/padded_batched_gemm_scale_softmax_gemm_xdl_fp16.cpp

```diff
@@ -0,0 +1,397 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2022, Advanced Micro Devices, Inc. All rights reserved.
+
+/*
+Gemm + Softmax + Gemm fused operation. Computes C_g_m_o = Softmax(A_g_m_k * B0_g_k_n) * B1_g_n_o
+                                                                  |-----------------|
+                                                                          Gemm0
+                                                          |-------------------------------------|
+                                                                          Gemm1
+*/
+
+#include <iostream>
+#include <numeric>
+#include <initializer_list>
+#include <cstdlib>
+
+#include "ck/ck.hpp"
+#include "ck/tensor_operation/gpu/device/gemm_specialization.hpp"
+#include "ck/tensor_operation/gpu/device/device_batched_gemm_softmax_gemm_xdl_cshuffle.hpp"
+#include "ck/tensor_operation/gpu/element/element_wise_operation.hpp"
+
+#include "ck/library/utility/check_err.hpp"
+#include "ck/library/utility/device_memory.hpp"
+#include "ck/library/utility/host_tensor.hpp"
+#include "ck/library/utility/host_tensor_generator.hpp"
+#include "ck/library/reference_tensor_operation/cpu/reference_batched_gemm.hpp"
+#include "ck/library/reference_tensor_operation/cpu/reference_softmax.hpp"
+
+template <ck::index_t... Is>
+using S = ck::Sequence<Is...>;
+
+using F16 = ck::half_t;
+using F32 = float;
+
+using Row = ck::tensor_layout::gemm::RowMajor;
+using Col = ck::tensor_layout::gemm::ColumnMajor;
+
+using PassThrough = ck::tensor_operation::element_wise::PassThrough;
+
+using ADataType        = F16;
+using B0DataType       = F16;
+using B1DataType       = F16;
+using AccDataType      = F32;
+using CShuffleDataType = F32;
+using CDataType        = F16;
+
+using ALayout  = Row;
+using B0Layout = Col;
+using B1Layout = Row;
+using CLayout  = Row;
+
+using AElementOp    = PassThrough;
+using B0ElementOp   = PassThrough;
+using Acc0ElementOp = ck::tensor_operation::element_wise::ScaleAndResetNaNToMinusInfinity;
```

Comment:
From average user's perspective, he/she has to do both right:
- Set GEMM N padding spec
- Set correct Acc0ElementOp becuase regular `element_wise::scale` won't handle N padding correctly

Can we handle `ResetNaNToMinusInfinity` transparently once N padding is on? I think those implementation details shouldn't be exposed to the interface

Reply:
As ```ScaleAndResetNaNToMinusInfinity```'s functionality scope contains ```scale```'s function in attention layer and it does not cause performance drop, we can set ```ScaleAndResetNaNToMinusInfinity``` as default ```Acc0ElementOp```. In that case, a user can get right and high perf result by only setting GEMM MN padding spec.

> Can we handle ResetNaNToMinusInfinity transparently once N padding is on?

Do you mean that we move ```Acc0ElementOp ``` from device level template to grid level?

Reply:
We should never require user to specify `Acc0ElementOp = ResetNaNToMinusInfinity` because users are not concerned with how we handle padding with NaNs and INFs. What I mean is to keep `Acc0ElementOp = scale` as it is, and leave ResetNaNToMinusInfinity details to gridwise level. To reiterate: we can and should do ResetNaNToMinusInfinity somewhere somehow once N padding is on.

And once we've done that, the difference between the two example code should be pretty minimal. We may consider reorg the code into what @poyenc did with https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/example/01_gemm/run_gemm_example.inc

Reply:
> To reiterate: we can and should do ResetNaNToMinusInfinity somewhere somehow once N padding is on.

Leaving resetting values to gridwise level means that we need to add an operation after ```Acc0ElementOp```. In condition of fuse the scaling and resetting together to max the performance, can you point out where?

Reply:
> And once we've done that, the difference between the two example code should be pretty minimal. We may consider reorg the code into what @poyenc did with https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/example/01_gemm/run_gemm_example.inc

This is nice to have, but not mandatory for now, since there is only fp16 example

> To reiterate: we can and should do ResetNaNToMinusInfinity somewhere somehow once N padding is on.

This is tricky. By definition this device_op does BMM+element_op+softmax+BMM+element_op, so element_op lke ```ResetNaNToMinusInfinity``` should be defined by the caller of this device_op. Please open an github issue on this @shaojiewang

Reply:
As discussed, let's defer this issue to the future.